### PR TITLE
Automatically remove orphaned pod's empty volumes

### LINKED ISSUE #60987: Orphaned pod found - but volume paths are still present on disk
**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG

**What happened**:
Kubelet is periodically going into an error state and causing errors with our storage layer (ceph, shared filesystem). Upon cleaning out the orphaned pod directory things eventually right themselves.
* Workaround: `rmdir /var/lib/kubelet/pods/*/volumes/*rook/*`

**What you expected to happen**:
Kubelet should intelligently deal with orphaned pods. Cleaning a stale directory manually should not be required.

**How to reproduce it (as minimally and precisely as possible)**:
Using rook-0.7.0 (this isn't a rook problem as far as I can tell but this is how we're reproducing):
kubectl create -f rook-operator.yaml
kubectl create -f rook-cluster.yaml
kubectl create -f rook-filesystem.yaml

Mount/write to the shared filesystem and monitor /var/log/messages for the following:
`
kubelet: E0309 16:46:30.429770    3112 kubelet_volumes.go:128] Orphaned pod "2815f27a-219b-11e8-8a2a-ec0d9a3a445a" found, but volume paths are still present on disk : There were a total of 1 errors similar to this. Turn up verbosity to see them.
`

**Anything else we need to know?**:
This looks identical to the following: https://github.com/kubernetes/kubernetes/issues/45464 but for a different plugin.

**Environment**:
- Kubernetes version (use `kubectl version`):
`
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T12:22:21Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T11:55:20Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
`

- Cloud provider or hardware configuration:
Bare-metal private cloud
- OS (e.g. from /etc/os-release):
Red Hat Enterprise Linux Server release 7.4 (Maipo)
- Kernel (e.g. `uname -a`):
Linux 4.4.115-1.el7.elrepo.x86_64 #1 SMP Sat Feb 3 20:11:41 EST 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
kubeadm

### PR DESCRIPTION
**Pull request by lorenz**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Handles the case where empty unmounted volumes (for example from a restart) are present and the pod does not get fully cleaned up due to this. This PR is pretty careful and will never delete anything if it appears to be mounted or contains data.
**Which issue(s) this PR fixes**:
Fixes #60987

**Special notes for your reviewer**:
I am assuming that this path is never concurrently executed with a pod creation. This appears to be correct since it is gated on the pod not being known to K8s. It should still be safe otherwise but it might lead to K8s hanging when cleaning up a broken volume.

```release-note
NONE
```

### CODE REVIEW

**Comment by lorenz** [comment]
Could anyone review this? It's been sitting untouched for 5 days.

@Random-Liu
@vishh

**Comment by lorenz** [comment]
It's been nearly two weeks and the PR is still sitting unreviewed. The original issue has ~36 upvotes so a lot of people are running into it and I'd like to get this into 1.14, so I'm mentioning some additional people from sig-node-reviewers. Please tell me if this is inappropriate.

/cc @dashpole @ derekwaynecarr @tallclair

**Comment by lorenz** [comment]
The single test failure seems to be a flaky test, it has nothing to do with the code being touched and the testgrid says it fails a lot.

**Comment by dashpole** [comment]
The storage folks will know the corner-cases around this
/assign @jingxu97

**Comment by lorenz** [comment]
/retest

I rebased on current master, hopefully fixes the issue with the failing test. Also fixed a logic issue detected in my test cluster.

**Comment by lorenz** [comment]
Bumping this again to still get it into 1.14. The rebased commit now passes all tests and just needs a quick review.

**Comment by lorenz** [comment]
Ping @jingxu97 
Trying to still get this into 1.14

**Comment by redbaron** [comment]
Orphaned volumes plague all of us.  This PR looks sane, it does what cluster operators (humans) do to recover: delete volume directory with rmdir. maybe it is not fixing underneath root cause, but it removes one more reason to babysit nodes.

@kubernetes/sig-node-pr-reviews

**Comment by lorenz** [comment]
@redbaron The main reason these "orphaned" volumes exist is because of reboots that happen uncleanly (crashes, timeouts, hardware issues, ...), a minority happens due to software or operator faults. For the reboots we need automatic cleanup anyways. As soon as they are orphaned, there is nothing to be done other than purging them in my opinion, regardless of why the volume is orphaned. Maybe it'd be worth it to export the number of cleanups as a metric.

**Comment by derekwaynecarr** [comment]
@gnufied @childsb can you take a look at this from a storage perspective?

**Comment by derekwaynecarr** [comment]
the code location for this is right in the kubelet, it should be safe assuming there are no other unknown race conditions.  I am mildly concerned that we could have an odd race condition in kubelet based on https://github.com/kubernetes/kubernetes/pull/72105 but it needs more investigation.

**Comment by gnufied** [comment]
@lorenz does this actually fixes rook issue you linked? AFAICT - this PR fixes a very narrow case of orphan pod directories which may or may not be applicable to rook. The underlying problem is - volume plugins need to support proper volume reconstruction when kubelet restarts by reading volume state from the node's disk. If they can't they are broken.

I am not opposed to this PR, but there are so many different reasons for orphan pod directories. I would like to know specific use case that this PR fixes.

**Comment by gnufied** [comment]
Why I am asking for more specific information about this bug is - as someone commented on rook issue, you could get:

```
 rmdir: failed to remove ‘/var/lib/kubelet/pods/7b383940-3cc7-11e8-a78b-b8ca3a70880c/volumes
/rook.io~rook/backups’: Device or resource busy```

So naturally this PR will not help fixing that particular issue. Also, since this PR only fixes issues associated with unmounted and empty volume directories left on the node , I would like to know which volume type you were using when this happened. In 1.10 - a change went in that fixed some of the issues around reconstruction (cc @jingxu97 ) but if it is still happening for certain volume types, I would also like to investigate those volume type and make sure that reconstruction path works for them (so as root cause is eliminated).

File: pkg/kubelet/kubelet_volumes.go [hunk 1]
```diff
@@ -127,8 +129,23 @@ func (kl *Kubelet) cleanupOrphanedPodDirs(pods []*v1.Pod, runningPods []*kubecon
...
-			continue
+			for _, volumePath := range volumePaths {
+				// IsNotLikelyMountPoint() not good enough here, volume mounts are frequently bind-mounted, also not on a hot path
+				isNotMounted, err := kl.mounter.IsNotMountPoint(volumePath)
+				if err != nil {
+					orphanVolumeErrors = append(orphanVolumeErrors, fmt.Errorf("Orphaned pod %q found, but unable to evaluate mountpoints: %v", uid, err))
+					continue PodLoop
+				}
+				if isNotMounted {
+					if err := os.Remove(volumePath); err != nil { // Attempt to remove volume, will fail if there is for example data in there or a permission problem
```
Comment (gnufied) [inline pkg/kubelet/kubelet_volumes.go]: if `volumePath` is also used in a subpath and subpath is not cleaned up yet, removing the parent directory first will cause subpath to be never cleaned up.

```
~> mkdir -p foo bind_foo
~> mount -o bind foo bind_foo
~> rm -rf foo
~> rm -rf bind_foo
rm: cannot remove 'bind_foo': Device or resource busy
````

cc @msau42

Comment (gnufied) [reply pkg/kubelet/kubelet_volumes.go]: Now that I think of, if subpath is unmounted we should be fine. But since we aren't cleaning volumes if it has subpaths(bind mounts) left on the disks, this change seems incompatible with that.

Comment (lorenz) [reply pkg/kubelet/kubelet_volumes.go]: This is mostly in there as a safety measure since the Remove command will fail if there are any files  left in there. It's not actually needed to remove anything.

**Comment by lorenz** [comment]
This whole thing is really two different issues. There is the issue that I'm solving which deals only with real orphaned volumes that are no longer mounted, which are caused mainly by unclean reboots and affect all volume types including secrets and configmaps. There is an additional issue with Rook where volumes don't get unmounted correctly at all (device or resource busy). This PR does nothing to fix that (but also doesn't make it worse) since that is not necessarily K8s fault anyways. My custom FlexVolume (now CSI) volume plugin never ran into any issues and my nodes now happily clean up all orphaned volumes with this PR.

File: pkg/kubelet/kubelet_volumes.go (hunk 1 above)
Comment (gnufied) [reply pkg/kubelet/kubelet_volumes.go]: The problem isn't that `Remove` function will fail. Problem is - something may have bind mounted this directory via a subpath to somewhere else and `isNotMounted` only checks if this directory is a mount point.

**Comment by gnufied** [comment]
ack. This code path has caused data loss in past and hence I will try to get more eyes to look at this. 

cc @kubernetes/sig-storage-bugs

File: pkg/kubelet/kubelet_volumes.go [hunk 2]
```diff
@@ -127,8 +129,23 @@ func (kl *Kubelet) cleanupOrphanedPodDirs(pods []*v1.Pod, runningPods []*kubecon
 			continue
 		}
 		if len(volumePaths) > 0 {
-			orphanVolumeErrors = append(orphanVolumeErrors, fmt.Errorf("Orphaned pod %q found, but volume paths are still present on disk", uid))
-			continue
+			for _, volumePath := range volumePaths {
```
Comment (gnufied) [inline pkg/kubelet/kubelet_volumes.go]: Can we move this code into a function of its own? Something like - `cleanupVolumePaths(volumePaths) error` . The idea is - if this function returns error then we can call continue and get rid of continue label which is bit hard to read.

**Comment by msau42** [comment]
On kubelet restart, reconstruction should now call UnmountVolume even if the plugin failed to properly ConstructVolumeSpec.  UnmountVolume does subpath cleanup first, and then pod directory cleanup.  So I don't think the cleanup orphaned pods path needs to handle this.

If we can consistently repro the issue, can we try to instrument the reconstruction logic to see if there is an issue there?

**Comment by lorenz** [comment]
@msau42 What is the minimum version for this behavior? And does that affect already orphaned volumes?

**Comment by gnufied** [comment]
I think the work mostly went in 1.10 but there were some follow up commits that fixed reconstruction bits.  Overall all fixes should be available in 1.11 and onwards. When a kubelet is shuts down uncleanly and then comes up - you may still see "orphan pod" message couple of times but soon the reconstruction logic should catch up and unmount the volume even if plugin did not reconstruct the volume spec properly. 

If this is still happening, it might be worth opening bugs that have specific logs.  You mentioned it is happening for your flexvolume plugin. Can you open a bug report and post some logs?

**Comment by zgfh** [comment]
hi，i have some problem.
my env:
kubernetes: 1.12.3
CSI : ceph-csi

log
```
Mar 21 14:16:18 dc1k8s01n01p.geniusafc.com kubelet[2713]: E0321 14:16:18.344986    2713 kubelet_volumes.go:140] Orphaned pod "177d81ca-3644-11e9-a997-801844f21628" found, but volume paths are still present on disk : There were a total of 3 errors similar to this. Turn up verbosity to see them.
Mar 21 14:16:20 dc1k8s01n01p.geniusafc.com kubelet[2713]: E0321 14:16:20.357821    2713 kubelet_volumes.go:140] Orphaned pod "177d81ca-3644-11e9-a997-801844f21628" found, but volume paths are still present on disk : There were a total of 3 errors similar to this. Turn up verbosity to see them.
```

i have to delete it by myself
```
rm -rf /var/lib/kubelet/pods/177d81ca-3644-11e9-a997-801844f21628/volumes/kubernetes.io~csi/pvc-*
```

**Comment by fredkan** [comment]
PR#68616 also for this issue: https://github.com/kubernetes/kubernetes/pull/68616
Maybe the "cleanupOrphanedPodDirs" be better deal with in storage plugin(csi, flexvolume);

**Comment by lorenz** [comment]
@fredkan I've looked at your PR, but I consider @jingxu97's point about doing potentially hanging operations in the main thread valid. I've specifically architected this not to perform any potenially hanging operations. At least on my nodes this is enough, I've never had any orphaned volumes again.

The question is still if this needs to happen at all or if this is 100% on FlexVolumes/CSI. I'd argue to merge it anyways since it will not negatively impact any future fixes, but as can be seen there is no consensus on that yet.

**Comment by lorenz** [comment]
@gnufied It's my own FlexVolume plugin. It's not so easy to reproduce, but I've done some digging and I think I've found a possible root cause:
`Feb 15 05:55:06 red kubelet-wrapper[2564]: W0215 05:55:06.329449    2564 reconciler.go:394] Could not construct volume information, cleanup the mounts. (pod.UID 11f2ec81-1222-11e9-91f9-74d4358c6859, volume.SpecName pvc-e74864be-dac9-11e8-aa41-74d4358c6859): no volume plugin matched`
This volume later becomes orphaned and is part of my FlexVolume.

@msau42 Seems hard to repro, but I can try. I've got at least one node which has orphaned volumes from 1.13.

**Comment by xichengliudui** [comment]
More and more people have this problem.
kubernetes: `1.12.3` also came along.
Ref: https://github.com/xichengliudui/public-repo/blob/master/logs-02.md

**Comment by xichengliudui** [comment]
There are some differences: after I removed the pod, not only did the empty directory appear, but the secret volume still exists.

**Comment by xichengliudui** [comment]
@lorenz ，These are the logs I threw, i don't know if I can help you.
Ref: https://github.com/xichengliudui/public-repo/blob/master/Orphaned-pod-bug.md

**Comment by jingxu97** [comment]
sorry to reply this late. I think we do have cases as @lorenz that volume dirs cannot be cleaned up due to crashes or other issues. Kubelet volume manager reconstruct will try to clean up mounts when kubelet restarts, but it might fail with different reasons. 

Pod is considered as Orphaned when pod object is deleted from API server and cannot be found from  container runtime cache. I think it should be safe to remove volume dirs if they are no longer mounted.

**Comment by msau42** [comment]
There are may be volume types (like Flex or CSI) that don't actually use mountpoints.  In which case deleting the volume dir can potentially delete data.

I think we should look into the root cause of why the volume teardown is not cleaning up correctly.

**Comment by jingxu97** [comment]
A couple of corner case that cleanupMount might not work during kubelet reconstruction process.
1. kubelet fails to read dir temporarily. Since reconstruction only happens once, it will not retry
2. pod is force deleted, dir is just unmounted but not removed. since kubelet is not restarting, it will not go through cleanupMount. But this seems not the concern of the issue this PR is trying to address. The main concern here is pod is stuck for cleaning up due to empty dir still exist?
3. UnmountVolume fails for some reason. There are multiple steps for unmounting volumes. If any step returns error, kubelet reconstruction process will not retry.

**Comment by lorenz** [comment]
@msau42 I've taken potential deletion of data into consideration, since it's a pretty obvious and arguably one of the most serious failure modes (data loss). I'm only issuing a single os.Remove() which is internally implemented as unlink() and rmdir(). Both calls cannot result in data loss since K8s makes sure the volume path is a directory at which point unlink() will fail and rmdir() will fail with ENOTEMPTY if there is anything left under it.

I do agree that better reconstruction is nice, but this will not solve already orphaned volumes. I know this will not solve every problem with every potential configuration (stuck mounts, ...), but it solves the issue strictly better than the current solution.

**Review by msau42 — COMMENTED** [review]
This seems fine to me to improve resilience to bugs and failures in the reconstruction logic and/or volume plugins.  It does muddle the separation of responsibilities between reconstruction and orphaned pod handle and will potentially mask bugs in reconstruction or the plugin.

cc @jsafrane who's also currently fixing bugs in reconstruction.

File: pkg/kubelet/kubelet_volumes.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fredkan) [reply pkg/kubelet/kubelet_volumes.go]: Agree, the subpath should be considered here.

**Comment by msau42** [comment]
Hm one more case to check is raw block. Volume path will be a symlink to a device in that case.  We need to carefully consider the safety of removing that symlink out of band

**Comment by lorenz** [comment]
> It does muddle the separation of responsibilities between reconstruction and orphaned pod handle and will potentially mask bugs in reconstruction or the plugin.

I would be onboard with logging orphan volume cleanups. That way we could still see these issues without spamming the log.

> Hm one more case to check is raw block. Volume path will be a symlink to a device in that case. We need to carefully consider the safety of removing that symlink out of band

Removing a symlink should never cause a safety issue, regardless if it's pointing at a directory or a block volume. Even for things like in-container mounts (which cannot be used in non-SYS_MOUNT containers, so are restricted to special things anyways) the symlinks get resolved at mount time and will not result in a broken or hung mount.

I'll address the other items discussed tomorrow.

**Comment by zhucan** [comment]
@lorenz @msau42 What version will solve this problem？
